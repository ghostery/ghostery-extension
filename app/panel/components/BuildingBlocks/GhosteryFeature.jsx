/**
 * Ghostery Feature Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2020 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import PropTypes from 'prop-types';
import ClassNames from 'classnames';
import Tooltip from '../../../shared-components/Tooltip';
import globals from '../../../../src/classes/Globals';

const { BLACKLISTED, WHITELISTED } = globals;

/**
 * @class Rendering and interaction for Ghostery feature button toggles
 * @memberof PanelBuildingBlocks
 */
class GhosteryFeature extends React.Component {
	constructor(props) {
		super(props);

		this.handleClick = this.handleClick.bind(this);
	}

	static _getButtonText(sitePolicy, showText, type) {
		if (!showText) {
			return '';
		}

		switch (type) {
			case 'trust':
				return (sitePolicy === WHITELISTED ?
					t('summary_trust_site_active') :
					t('summary_trust_site'));
			case 'restrict':
				return (sitePolicy === BLACKLISTED ?
					t('summary_restrict_site_active') :
					t('summary_restrict_site'));
			default:
				return 'Check button type you are passing to GhosteryFeature for typos and make sure it is being handled by getButtonText';
		}
	}

	static _getTooltipText(sitePolicy, type) {
		switch (type) {
			case 'trust':
				return (sitePolicy === WHITELISTED ?
					t('tooltip_trust_on') :
					t('tooltip_trust'));
			case 'restrict':
				return (sitePolicy === BLACKLISTED ?
					t('tooltip_restrict_on') :
					t('tooltip_restrict'));
			default:
				return 'Check button type you are passing to GhosteryFeature for typos and make sure it is being handled by getTooltipText';
		}
	}

	static _isFeatureActive(type, sitePolicy) {
		switch (type) {
			case 'trust':
				return sitePolicy === WHITELISTED;
			case 'restrict':
				return sitePolicy === BLACKLISTED;
			default:
				return false;
		}
	}

	/**
	 * Handles user click on the Ghostery Feature button
	 */
	handleClick() {
		const { blockingPausedOrDisabled, handleClick, type } = this.props;
		if (blockingPausedOrDisabled) {
			return;
		}

		handleClick(type);
	}

	render() {
		const {
			blockingPausedOrDisabled,
			narrow,
			short,
			showText,
			sitePolicy,
			tooltipPosition,
			type
		} = this.props;

		const active = GhosteryFeature._isFeatureActive(type, sitePolicy);
		// TODO Foundation dependency: button
		const ghosteryFeatureClassNames = ClassNames(
			'button',
			'g-tooltip',
			'GhosteryFeatureButton',
			{
				'GhosteryFeatureButton--normal': !narrow && !short,
				'GhosteryFeatureButton--short': short,
				'GhosteryFeatureButton--narrow': narrow,
				'GhosteryFeatureButton--inactive': !active,
				'GhosteryFeatureButton--active': active,
				trust: type === 'trust',
				restrict: type === 'restrict',
				clickable: !blockingPausedOrDisabled,
				'not-clickable': blockingPausedOrDisabled,
			}
		);

		// TODO Foundation dependency: flex-container, align-center-middle
		return (
			<div className={ghosteryFeatureClassNames} onClick={this.handleClick}>
				<span className="flex-container align-center-middle full-height">
					<span className="GhosteryFeatureButton__text">
						{GhosteryFeature._getButtonText(sitePolicy, showText, type)}
					</span>
				</span>
				<Tooltip body={GhosteryFeature._getTooltipText(sitePolicy, type)} position={tooltipPosition} />
			</div>
		);
	}
}

GhosteryFeature.propTypes = {
	handleClick: PropTypes.func.isRequired,
	type: PropTypes.oneOf(['trust', 'restrict']).isRequired,
	sitePolicy: PropTypes.oneOf([false, 1, 2]),
	blockingPausedOrDisabled: PropTypes.bool.isRequired,
	showText: PropTypes.bool.isRequired,
	tooltipPosition: PropTypes.string.isRequired,
	short: PropTypes.bool.isRequired,
	narrow: PropTypes.bool.isRequired,
};

GhosteryFeature.defaultProps = {
	sitePolicy: false,
};

export default GhosteryFeature;
