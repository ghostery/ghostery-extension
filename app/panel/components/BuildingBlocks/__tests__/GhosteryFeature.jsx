/**
 * Ghostery Feature Test Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2020 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import GhosteryFeature from '../GhosteryFeature';

// Fake the translation function to only return the translation key
global.t = function(str) {
	return str;
};

// Fake the Tooltip implementation
jest.mock('../../../../shared-components/Tooltip');

describe('app/panel/components/GhosteryFeature.jsx', () => {
	describe('Snapshot tests with react-test-renderer', () => {
		test('GhosteryFeature is rendered correctly props are falsy', () => {
			const component = renderer.create(
				<div>
					<GhosteryFeature
						handleClick={() => {}}
						type="trust"
						sitePolicy={false}
						blockingPausedOrDisabled={false}
						showText={false}
						tooltipPosition=""
						short={false}
						narrow={false}
					/>
					<GhosteryFeature
						handleClick={() => {}}
						type="restrict"
						sitePolicy={false}
						blockingPausedOrDisabled={false}
						showText={false}
						tooltipPosition=""
						short={false}
						narrow={false}
					/>
				</div>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('GhosteryFeature is rendered correctly some props are truthy', () => {
			const component = renderer.create(
				<div>
					<GhosteryFeature
						handleClick={() => {}}
						type="trust"
						sitePolicy={2}
						blockingPausedOrDisabled={false}
						showText
						tooltipPosition="right"
						short
						narrow={false}
					/>
					<GhosteryFeature
						handleClick={() => {}}
						type="restrict"
						sitePolicy={2}
						blockingPausedOrDisabled={false}
						showText
						tooltipPosition="right"
						short
						narrow={false}
					/>
				</div>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('GhosteryFeature is rendered correctly props are truthy', () => {
			const component = renderer.create(
				<div>
					<GhosteryFeature
						handleClick={() => {}}
						type="trust"
						sitePolicy={1}
						blockingPausedOrDisabled
						showText
						tooltipPosition="top"
						short
						narrow
					/>
					<GhosteryFeature
						handleClick={() => {}}
						type="restrict"
						sitePolicy={1}
						blockingPausedOrDisabled
						showText
						tooltipPosition="top"
						short
						narrow
					/>
				</div>
			).toJSON();
			expect(component).toMatchSnapshot();
		});
	});

	describe('Shallow snapshot tests rendered with Enzyme', () => {
		test('GhosteryFeature has the correct class names', () => {
			const component = shallow(
				<GhosteryFeature
					handleClick={() => {}}
					type="restrict"
					sitePolicy={false}
					blockingPausedOrDisabled={false}
					showText={false}
					tooltipPosition=""
					short={false}
					narrow={false}
				/>
			);
			expect(component.find('.GhosteryFeatureButton.restrict').length).toBe(1);
			expect(component.find('.GhosteryFeatureButton.trust').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton.not-clickable').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton.clickable').length).toBe(1);
			expect(component.find('.GhosteryFeatureButton--normal').length).toBe(1);
			expect(component.find('.GhosteryFeatureButton--short').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--narrow').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--inactive').length).toBe(1);
			expect(component.find('.GhosteryFeatureButton--active').length).toBe(0);

			component.setProps({ narrow: true, sitePolicy: 1 });
			expect(component.find('.GhosteryFeatureButton--normal').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--short').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--narrow').length).toBe(1);
			expect(component.find('.GhosteryFeatureButton--inactive').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--active').length).toBe(1);

			component.setProps({ short: true, type: 'trust' });
			expect(component.find('.GhosteryFeatureButton.restrict').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton.trust').length).toBe(1);
			expect(component.find('.GhosteryFeatureButton--normal').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--short').length).toBe(1);
			expect(component.find('.GhosteryFeatureButton--narrow').length).toBe(1);
			expect(component.find('.GhosteryFeatureButton--inactive').length).toBe(1);
			expect(component.find('.GhosteryFeatureButton--active').length).toBe(0);

			component.setProps({ narrow: false, sitePolicy: 2 });
			expect(component.find('.GhosteryFeatureButton--normal').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--short').length).toBe(1);
			expect(component.find('.GhosteryFeatureButton--narrow').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--inactive').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--active').length).toBe(1);

			component.setProps({ short: false, blockingPausedOrDisabled: true });
			expect(component.find('.GhosteryFeatureButton.not-clickable').length).toBe(1);
			expect(component.find('.GhosteryFeatureButton.clickable').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--normal').length).toBe(1);
			expect(component.find('.GhosteryFeatureButton--short').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--narrow').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--inactive').length).toBe(0);
			expect(component.find('.GhosteryFeatureButton--active').length).toBe(1);
		});

		test('GhosteryFeature handles clicks correctly', () => {
			const handleClick = jest.fn();
			const component = shallow(
				<GhosteryFeature
					handleClick={handleClick}
					type="restrict"
					sitePolicy={false}
					blockingPausedOrDisabled
					showText={false}
					tooltipPosition=""
					short={false}
					narrow={false}
				/>
			);

			expect(handleClick.mock.calls.length).toBe(0);
			component.find('.GhosteryFeatureButton').simulate('click');
			expect(handleClick.mock.calls.length).toBe(0);

			component.setProps({ blockingPausedOrDisabled: false });
			component.find('.GhosteryFeatureButton').simulate('click');
			expect(handleClick.mock.calls.length).toBe(1);
		});
	});
});
