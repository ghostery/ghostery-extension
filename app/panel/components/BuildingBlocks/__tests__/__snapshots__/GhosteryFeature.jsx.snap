// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/panel/components/GhosteryFeature.jsx Snapshot tests with react-test-renderer GhosteryFeature is rendered correctly props are falsy 1`] = `
<div>
  <div
    className="button g-tooltip GhosteryFeatureButton GhosteryFeatureButton--normal GhosteryFeatureButton--inactive trust clickable"
    onClick={[Function]}
  >
    <span
      className="flex-container align-center-middle full-height"
    >
      <span
        className="GhosteryFeatureButton__text"
      >
        
      </span>
    </span>
  </div>
  <div
    className="button g-tooltip GhosteryFeatureButton GhosteryFeatureButton--normal GhosteryFeatureButton--inactive restrict clickable"
    onClick={[Function]}
  >
    <span
      className="flex-container align-center-middle full-height"
    >
      <span
        className="GhosteryFeatureButton__text"
      >
        
      </span>
    </span>
  </div>
</div>
`;

exports[`app/panel/components/GhosteryFeature.jsx Snapshot tests with react-test-renderer GhosteryFeature is rendered correctly props are truthy 1`] = `
<div>
  <div
    className="button g-tooltip GhosteryFeatureButton GhosteryFeatureButton--short GhosteryFeatureButton--narrow GhosteryFeatureButton--inactive trust not-clickable"
    onClick={[Function]}
  >
    <span
      className="flex-container align-center-middle full-height"
    >
      <span
        className="GhosteryFeatureButton__text"
      >
        summary_trust_site
      </span>
    </span>
  </div>
  <div
    className="button g-tooltip GhosteryFeatureButton GhosteryFeatureButton--short GhosteryFeatureButton--narrow GhosteryFeatureButton--active restrict not-clickable"
    onClick={[Function]}
  >
    <span
      className="flex-container align-center-middle full-height"
    >
      <span
        className="GhosteryFeatureButton__text"
      >
        summary_restrict_site_active
      </span>
    </span>
  </div>
</div>
`;

exports[`app/panel/components/GhosteryFeature.jsx Snapshot tests with react-test-renderer GhosteryFeature is rendered correctly some props are truthy 1`] = `
<div>
  <div
    className="button g-tooltip GhosteryFeatureButton GhosteryFeatureButton--short GhosteryFeatureButton--active trust clickable"
    onClick={[Function]}
  >
    <span
      className="flex-container align-center-middle full-height"
    >
      <span
        className="GhosteryFeatureButton__text"
      >
        summary_trust_site_active
      </span>
    </span>
  </div>
  <div
    className="button g-tooltip GhosteryFeatureButton GhosteryFeatureButton--short GhosteryFeatureButton--inactive restrict clickable"
    onClick={[Function]}
  >
    <span
      className="flex-container align-center-middle full-height"
    >
      <span
        className="GhosteryFeatureButton__text"
      >
        summary_restrict_site
      </span>
    </span>
  </div>
</div>
`;
